metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py

```diff
@@ -237,6 +270,343 @@ def create(cls, config_dict: Dict, ctx: PipelineContext) -> "SQLServerSource":
+                            SET
+                              referenced_object_type = ri.referenced_type,
+                              referenced_schema_name = ri.referenced_schema,
+                              referenced_entity_name = ri.referenced_name
+                            FROM #ProceduresDependencies pd
+                            INNER JOIN @referenced_info ri
+                                ON ri.id = pd.id
+                            WHERE CURRENT OF proc_depend_cursor;
+
+                            SET @referenced_object_type = (
+                                SELECT
+                                    referenced_type
+                                FROM @referenced_info
+                                WHERE id = @id
+                                );
+                        END TRY
+                        BEGIN CATCH
+                            PRINT ERROR_MESSAGE()
+                        END CATCH;
+                    END;
+
+                IF @referenced_object_type = 'SN'
+                  BEGIN
+                      BEGIN TRY
+                          SET @SQL  = 'SELECT
+                                      ' + CAST(@id as NVARCHAR(10)) + '
+                                        , syn_obj.name
+                                        , syn_obj.type
+                                        , syn_obj.object_id
+                                      FROM ' + COALESCE(QUOTENAME(@referenced_database_name), '') + '.sys.synonyms AS synon
+                                      INNER JOIN ' + COALESCE(QUOTENAME(@referenced_database_name), '') + '.sys.objects  AS syn_obj
+                                            ON OBJECT_ID(synon.base_object_name) = syn_obj.object_id
+                                      WHERE synon.object_id = ' + CAST(@referenced_id AS NVARCHAR(10)) + ';';
+                            PRINT @SQL;
+                          INSERT INTO @synonym_info(
+                              id
+                            , referenced_name
+                            , referenced_type
+                            , referenced_object_id
+                              )
+                          EXEC sp_executesql @SQL;
+
+                          UPDATE #ProceduresDependencies
+                          SET
+                              referenced_object_type = si.referenced_type,
+                              referenced_entity_name = si.referenced_name,
+                              referenced_id = si.referenced_object_id
+                          FROM #ProceduresDependencies pd
+                          INNER JOIN @synonym_info si
+                              ON si.id = pd.id
+                          WHERE CURRENT OF proc_depend_cursor;
+                      END TRY
+                      BEGIN CATCH
+                          PRINT ERROR_MESSAGE()
+                      END CATCH;
+                  END;
+
+                FETCH NEXT FROM proc_depend_cursor INTO
```

Comment:
why does this need to be a procedure, rather than a more simple sql statement?

It's also a bit difficult for me to understand all of the logic in here - some comments and explanation would useful